x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -459,10 +459,93 @@ public void analyzedPlan(
+            listener.onResponse(preAnalysisResult);
+        }
+    }
+
+    private void preAnalyzeSubqueryIndex(
+        PreAnalyzer.PreAnalysis preAnalysis,
+        IndexPattern subqueryIndexPattern,
+        PreAnalysisResult result,
+        EsqlExecutionInfo executionInfo,
+        ActionListener<PreAnalysisResult> listener
+    ) {
+        assert ThreadPool.assertCurrentThreadPool(
+            ThreadPool.Names.SEARCH,
+            ThreadPool.Names.SEARCH_COORDINATION,
+            ThreadPool.Names.SYSTEM_READ
+        );
+        if (subqueryIndexPattern != null) {
+            /*
+             * TODO subqueries with remote clusters need to be tested.
+             * Subquery index pattern can contain remote clusters, which need to be
+             * resolved against the available clusters. The input executionInfo is built for
+             *  the main index pattern, not for subqueries. Create an index pattern with
+             *  remote clusters for subquery's index pattern, make a copy of the main
+             * EsqlExecutionInfo for this subquery, and reuse the existing API to build
+             * the subqueryIndexExpression.
+             */
+            String indexExpressionToResolve = subqueryIndexExpression(executionInfo, subqueryIndexPattern);
+            if (indexExpressionToResolve.isEmpty()) {
+                listener.onResponse(
+                    result.addSubqueryIndexResolution(subqueryIndexPattern.indexPattern(), IndexResolution.invalid("[none available]"))
+                );
+                return;
+            }
+            // time-series index is not supported in subqueries yet, the grammar does not allow it
+            indexResolver.resolveAsMergedMapping(
+                indexExpressionToResolve,
+                result.fieldNames,
+                null,
+                false,
+                false,
+                preAnalysis.supportsDenseVector(),
+                listener.delegateFailure((l, indexResolution) -> {
+                    l.onResponse(result.addSubqueryIndexResolution(subqueryIndexPattern.indexPattern(), indexResolution));
+                })
+            );
+        } else {
+            // occurs when dealing with local relations (row a = 1)
+            listener.onResponse(result.withIndices(IndexResolution.invalid("[none specified]")));
+        }
+    }
+
+    private String subqueryIndexExpression(EsqlExecutionInfo mainExecutionInfo, IndexPattern subqueryIndexPattern) {
+        // Clone mainInfo (assuming a copy constructor or similar method exists)
+        EsqlExecutionInfo subqueryExecutionInfo = new EsqlExecutionInfo(
+            mainExecutionInfo.skipOnFailurePredicate(),
+            mainExecutionInfo.includeCCSMetadata()
+        );
+        EsqlCCSUtils.initCrossClusterState(indicesExpressionGrouper, verifier.licenseState(), subqueryIndexPattern, subqueryExecutionInfo);
+
+        return EsqlCCSUtils.createIndexExpressionFromAvailableClusters(subqueryExecutionInfo);
```

Comment:
The way we resolve expressions is going to change soon: we are going to resolve IndexPattern directly.
Lets coordinate how to integrate that change.

Reply:
If we have multiple `FROM` patterns, it will probably make sense to create a common field-caps call? But having separate ones may be ok for starters too.

We should be careful with `subqueryExecutionInfo` though - for example, if some clusters are marked as skipped there, should they also be skipped in the main query? I am concerned we could get inconsistent results where one of the subqueries skips one set of clusters, another skips other clusters, and main query would skip third set of clusters, and it would be impossible to understand what happened in the result (and also report what happened, since we only have one `_clusters` in the output). Would be nice to document what should happen in such case.